### PR TITLE
ERC20WithPermit test improvements

### PR DESCRIPTION
**Pull request by nkuba**
This PR is a follow-up for https://github.com/thesis/solidity-contracts/pull/3 to improve unit tests.

### CODE REVIEW

File: test/token/ERC20WithPermit.test.js [hunk 1]
```diff
@@ -114,8 +134,57 @@ describe("ERC20WithPermit", () => {
...
+            await token
+              .connect(initialHolder)
+              .transfer(recipient.address, amount2)
+
+            expect(await token.balanceOf(initialHolder.address)).to.equal(
+              initialHolderBalance.sub(amount1).sub(amount2)
+            )
+
+            expect(await token.balanceOf(recipient.address)).to.equal(
+              amount1.add(amount2)
```
Comment (tomaszslabon) [inline test/token/ERC20WithPermit.test.js:180]: Test

File: test/token/ERC20WithPermit.test.js [hunk 2]
```diff
@@ -114,8 +134,57 @@ describe("ERC20WithPermit", () => {
...
+        it("should emit a transfer event", async () => {
+          const tx = await token
+            .connect(initialHolder)
+            .transfer(recipient.address, amount)
+
+          await expect(tx)
+            .to.emit(token, "Transfer")
+            .withArgs(initialHolder.address, recipient.address, amount)
+        })
+      })
```
Comment (tomaszslabon) [inline test/token/ERC20WithPermit.test.js:159]: Consider calling `transfer` just once in the whole context (for example in `beforeEach`):
```
context("when the sender transfers part of their tokens", () => {
  const amount = to1e18(60)
  let tx
  beforeEach(async () => {
    tx = await token
      .connect(initialHolder)
      .transfer(recipient.address, amount)
  })

  it("should transfer the requested amount", async () => {
    expect(await token.balanceOf(initialHolder.address)).to.equal(
      initialHolderBalance.sub(amount)
    )

    expect(await token.balanceOf(recipient.address)).to.equal(amount)
  })

  it("should emit a transfer event", async () => {
    await expect(tx)
      .to.emit(token, "Transfer")
      .withArgs(initialHolder.address, recipient.address, amount)
  })
})
```

I saw this pattern in most unit tests, for example for [`mint` tests](https://github.com/thesis/solidity-contracts/blob/431fb388b9aee13869ebf69d6967d1e89b64cd19/test/token/ERC20WithPermit.test.js#L699).

Comment (nkuba) [reply test/token/ERC20WithPermit.test.js]: 0ef7ee3cc8758b9613e107625494da8b4d4ce51f

File: test/token/ERC20WithPermit.test.js [hunk 3]
```diff
@@ -95,16 +112,20 @@ describe("ERC20WithPermit", () => {
...
         expect(await token.balanceOf(initialHolder.address)).to.equal(
-          initialSupply
+          initialHolderBalance
         )
       })
     })
   })
 
   describe("transfer", () => {
+    afterEach("total supply remains unchanged", async () => {
```
Comment (mhluongo) [inline test/token/ERC20WithPermit.test.js:122]: :+1:

File: test/token/ERC20WithPermit.test.js [hunk 4]
```diff
@@ -164,76 +233,207 @@ describe("ERC20WithPermit", () => {
...
       })
     })
   })
 
   describe("transferFrom", () => {
+    afterEach("total supply remains unchanged", async () => {
+      expect(await token.totalSupply()).to.equal(initialSupply)
+    })
+
     context("when the token owner is not the zero address", () => {
```
[the remark this answers is not in the document]
Comment (mhluongo) [reply test/token/ERC20WithPermit.test.js:273]: We should, but let's do it in another PR :)

File: test/token/ERC20WithPermit.test.js [hunk 5]
```diff
@@ -793,25 +968,100 @@ describe("ERC20WithPermit", () => {
...
+          const deadline = tomorrow
+          const signature = await getApproval(
+            permittingHolderBalance,
+            recipient.address,
+            deadline
+          )
+
+          const malleableS = ethers.BigNumber.from(
+            "0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0"
+          ).add(1)
```
Comment (mhluongo) [inline test/token/ERC20WithPermit.test.js:1028]: I think I know how you got here and it shouldn't block merge... but this is a good spot to include a link to learn more about malleability and explain this magic string

**Review by mhluongo — APPROVED** [review]
Really appreciate the attention to detail here @nkuba. Left a couple comments but I'm merging this in the meantime.

**Pull request merged**